ccl/sqlproxyccl: TestConnectionMigration failed

ccl/sqlproxyccl.TestConnectionMigration [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=6336563&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=6336563&tab=artifacts#/) on release-22.1 @ [443dd822332bb6487c949efbd9a4b700693850d4](https://github.com/cockroachdb/cockroach/commits/443dd822332bb6487c949efbd9a4b700693850d4):


```
=== RUN   TestConnectionMigration
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/fe188f163e90a7ffb92e92cbaea22b0c/logTestConnectionMigration1739277136
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestConnectionMigration
    proxy_handler_test.go:1314: -- test log scope end --
--- FAIL: TestConnectionMigration (10.23s)
=== RUN   TestConnectionMigration/successful/normal_transfer
    proxy_handler_test.go:1062: 
        	Error Trace:	proxy_handler_test.go:1062
        	Error:      	Received unexpected error:
        	            	transfer cannot be started
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/conn_migration.go:114
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6498
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6475
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:238
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1581
        	            	Wraps: (2) transfer cannot be started
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestConnectionMigration/successful/normal_transfer
        --- FAIL: TestConnectionMigration/successful/normal_transfer (0.03s)
=== RUN   TestConnectionMigration/successful
    --- FAIL: TestConnectionMigration/successful (3.27s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
/cc @cockroachdb/sql-experience @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConnectionMigration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-19287

ccl/sqlproxyccl.TestConnectionMigration [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=7687249&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=7687249&tab=artifacts#/) on release-22.1 @ [6b3f19ae0f6f47bb606d47a07fd5e0f8a80e4fe8](https://github.com/cockroachdb/cockroach/commits/6b3f19ae0f6f47bb606d47a07fd5e0f8a80e4fe8):


```
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1581
        	            	Wraps: (2) transfer cannot be started
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestConnectionMigration/successful/normal_transfer
        --- FAIL: TestConnectionMigration/successful/normal_transfer (0.00s)
=== RUN   TestConnectionMigration/successful/failed_transfers_with_tx
    proxy_handler_test.go:1140: 
        	Error Trace:	proxy_handler_test.go:1140
        	            				tx.go:160
        	            				common.go:70
        	            				tx.go:160
        	            				proxy_handler_test.go:1133
        	Error:      	Expect "transfer cannot be started" to match "cannot serialize"
        	Test:       	TestConnectionMigration/successful/failed_transfers_with_tx
    proxy_handler_test.go:1154: 
        	Error Trace:	proxy_handler_test.go:1154
        	Error:      	Received unexpected error:
        	            	non-serialization error: transfer cannot be started
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.TestConnectionMigration.func2.5.1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/proxy_handler_test.go:1141
        	            	  | github.com/cockroachdb/cockroach-go/v2/crdb.ExecuteTx.func1
        	            	  | 	github.com/cockroachdb/cockroach-go/v2/crdb/external/com_github_cockroachdb_cockroach_go_v2/crdb/tx.go:160
        	            	  | github.com/cockroachdb/cockroach-go/v2/crdb.ExecuteInTx
        	            	  | 	github.com/cockroachdb/cockroach-go/v2/crdb/external/com_github_cockroachdb_cockroach_go_v2/crdb/common.go:70
        	            	  | github.com/cockroachdb/cockroach-go/v2/crdb.ExecuteTx
        	            	  | 	github.com/cockroachdb/cockroach-go/v2/crdb/external/com_github_cockroachdb_cockroach_go_v2/crdb/tx.go:160
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.TestConnectionMigration.func2.5
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/proxy_handler_test.go:1133
        	            	  | [...repeated from below...]
        	            	Wraps: (2) non-serialization error
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/conn_migration.go:114
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6498
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6475
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:238
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1581
        	            	Wraps: (4) transfer cannot be started
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
        	Test:       	TestConnectionMigration/successful/failed_transfers_with_tx
        --- FAIL: TestConnectionMigration/successful/failed_transfers_with_tx (0.00s)
=== RUN   TestConnectionMigration/successful
    --- FAIL: TestConnectionMigration/successful (3.11s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConnectionMigration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/sqlproxyccl/conn_migration.go
+++ b/pkg/ccl/sqlproxyccl/conn_migration.go
@@ -111,6 +111,9 @@ func (f *forwarder) tryBeginTransfer() (started bool, cleanupFn func()) {
 	}
 }
 
+// errTransferCannotStart is an error that indicates that the transfer cannot be
+// started (e.g. transfer already in progress, or we're not at a safe transfer
+// point). The caller should retry the transfer again if necessary.
 var errTransferCannotStart = errors.New("transfer cannot be started")
 
 // TransferConnection attempts a best-effort connection migration to an

--- a/pkg/ccl/sqlproxyccl/proxy_handler_test.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler_test.go
@@ -1397,6 +1397,24 @@ func TestConnectionMigration(t *testing.T) {
 			proxy.metrics.ConnMigrationTransferResponseMessageSize.TotalCount())
 	}
 
+	transferConnWithRetries := func(t *testing.T, f *forwarder) error {
+		t.Helper()
+
+		var nonRetriableErrSeen bool
+		err := testutils.SucceedsSoonError(func() error {
+			err := f.TransferConnection()
+			if err == nil {
+				return nil
+			}
+			if !errors.Is(err, errTransferCannotStart) {
+				nonRetriableErrSeen = true
+			}
+			return err
+		})
+		require.False(t, nonRetriableErrSeen)
+		return err
+	}
+
 	// Test that connection transfers are successful. Note that if one sub-test
 	// fails, the remaining will fail as well since they all use the same
 	// forwarder instance.
@@ -1452,7 +1470,7 @@ func TestConnectionMigration(t *testing.T) {
 			require.NoError(t, err)
 
 			// Show that we get alternating SQL pods when we transfer.
-			require.NoError(t, f.TransferConnection())
+			require.NoError(t, transferConnWithRetries(t, f))
 			require.Equal(t, int64(1), f.metrics.ConnMigrationSuccessCount.Count())
 			require.Equal(t, tenant2.SQLAddr(), queryAddr(tCtx, t, db))
 
@@ -1463,7 +1481,7 @@ func TestConnectionMigration(t *testing.T) {
 			_, err = db.Exec("SET application_name = 'bar'")
 			require.NoError(t, err)
 
-			require.NoError(t, f.TransferConnection())
+			require.NoError(t, transferConnWithRetries(t, f))
 			require.Equal(t, int64(2), f.metrics.ConnMigrationSuccessCount.Count())
 			require.Equal(t, tenant1.SQLAddr(), queryAddr(tCtx, t, db))
 
@@ -1525,10 +1543,14 @@ func TestConnectionMigration(t *testing.T) {
 
 			err = crdb.ExecuteTx(tCtx, db, nil /* txopts */, func(tx *gosql.Tx) error {
 				// Run multiple times to ensure that connection isn't closed.
-				for i := 0; i < 5; i++ {
+				for i := 0; i < 5; {
 					err := f.TransferConnection()
 					if err == nil {
 						return errors.New("no error")
+					}
+					// Retry again if the transfer cannot be started.
+					if errors.Is(err, errTransferCannotStart) {
+						continue
 					}
 					if !assert.Regexp(t, "cannot serialize", err.Error()) {
 						return errors.Wrap(err, "non-serialization error")
@@ -1541,6 +1563,7 @@ func TestConnectionMigration(t *testing.T) {
 							addr,
 						)
 					}
+					i++
 				}
 				return nil
 			})
@@ -1554,7 +1577,7 @@ func TestConnectionMigration(t *testing.T) {
 			require.Equal(t, int64(0), f.metrics.ConnMigrationErrorFatalCount.Count())
 
 			// Once the transaction is closed, transfers should work.
-			require.NoError(t, f.TransferConnection())
+			require.NoError(t, transferConnWithRetries(t, f))
 			require.NotEqual(t, initAddr, queryAddr(tCtx, t, db))
 			require.Nil(t, f.ctx.Err())
 			require.Equal(t, initSuccessCount+1, f.metrics.ConnMigrationSuccessCount.Count())

